docs/features/string-literals-data-section.md

```diff
@@ -0,0 +1,201 @@
+
+## Eligible string literals
+
+A string literal is emitted with the utf8 encoding if and only if all of these are true:
+- it can be encoded as UTF-8, i.e., the following call does not throw:
+  ```cs
+  new System.Text.Encoding.UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true).GetBytes(literal)
+  ```
+- its `Length` is greater than the threshold specified by [the feature flag](#configuration).
+
+## Emit
+
+The standard string literal emit strategy emits `ldstr "string"`.
+That puts the string literal on the UserString heap of the PE file.
+
+The utf8 string literal encoding emit strategy emits `ldsfld` of a field in a generated class instead.
+
+For every string literal, a unique internal static class is generated which:
+- has name composed of `<S>` followed by a hex-encoded XXH128 hash of the string,
+- lives in the global namespace (for consistency with other compiler-generated types),
+- has one `.data` field which is emitted the same way as for [explicit u8 string literals][u8-literals],
+- has one `string` field which is initialized in a static constructor of the class.
+
+The initialization calls `<PrivateImplementationDetails>.BytesToString` helper which in turn calls `Encoding.UTF8.GetBytes`.
+This is an optimization so each of the generated class static constructors is slightly smaller in IL size.
+These size savings can add up since one class is generated per one eligible string literal.
+
+The following example demonstrates the code generated for string literal `"Hello."`.
+
+```cs
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+
+[CompilerGenerated]
+internal static class <S>F6D59F18520336E6E57244B792249482
+{
+    internal static readonly <PrivateImplementationDetails>.__StaticArrayInitTypeSize=6 f = /* IL: data(48 65 6C 6C 6F 2E) */;
+
+    internal static readonly string s;
+
+    unsafe static <S>F6D59F18520336E6E57244B792249482()
+    {
+        s = <PrivateImplementationDetails>.BytesToString((byte*)Unsafe.AsPointer(ref f), 6);
+    }
+}
+
+[CompilerGenerated]
+internal static class <PrivateImplementationDetails>
+{
+    [StructLayout(LayoutKind.Explicit, Pack = 1, Size = 6)]
+    internal struct __StaticArrayInitTypeSize=6
+    {
+    }
+
+    internal unsafe static string BytesToString(byte* bytes, int length)
+    {
+        return Encoding.UTF8.GetString(bytes, length);
+    }
+}
+```
+
+### ILVerify and PEVerify
+
+The actual generated code does not call `(byte*)Unsafe.AsPointer`.
+It simply emits `ldsflda` of the data field.
+That works but does not pass IL/PE verification.
+
+## Runtime
+
+Throughput of `ldstr` vs `ldsfld` is very similar (both result in one or two move instructions).
+
+In the `ldsfld` emit strategy, the `string` instances won't ever be collected by the GC once the generated class is initialized.
```

Comment:
I would speculate that users that hit User-String heap limit might prefer a behavior that allows these strings to be collected. This a trade-off that is often made, use less memory at the price of a small slowdown. For example, in our test assemblies, each unit-test is executed once and the rate of sharing of big literals between unit-tests is probably low. Therefore, it might be beneficial to not hold onto the strings for the entire duration of a test-assembly run, thus avoiding potential out of memory situations.

Reply:
I will mention this as future work. There are some trade-offs (the string loading is a more complicated IL than just `ldsfld`), so it's not clear this should be implemented as part of the first iteration.

Reply:
> I will mention this as future work. 

I do not see this mentioned. If you were referring to "Weak references" section, I do not think this is equivalent because use of weak references is not required to make strings collectable. Also, I think this is more an alternative rather than a "future work".